core/preconditioner/jacobi.hpp

```diff
@@ -0,0 +1,512 @@
+         * value of `storage_optimization.is_block_wise`, the
+         * `storage_optimization.block_wise` property will return an array of
+         * precisions used for each block. All values set to
+         * `precision_reduction::autodetect()` will be replaced with the value
+         * representing the precision used for the corresponding block.
+         * If the non-adaptive version of Jacobi is used, the
+         * `storage_optimization.block_wise` Array will be empty.
+         */
+        storage_optimization_type GKO_FACTORY_PARAMETER(
+            storage_optimization, precision_reduction(0, 0));
+    };
+    GKO_ENABLE_LIN_OP_FACTORY(Jacobi, parameters, Factory);
+    GKO_ENABLE_BUILD_METHOD(Factory);
+
+protected:
+    /**
+     * Creates an empty Jacobi preconditioner.
+     *
+     * @param exec  the executor this object is assigned to
+     */
+    explicit Jacobi(std::shared_ptr<const Executor> exec)
+        : EnableLinOp<Jacobi>(exec), blocks_(exec)
+    {
+        parameters_.block_pointers.set_executor(exec);
+        parameters_.storage_optimization.block_wise.set_executor(exec);
+    }
+
+    /**
+     * Creates a Jacobi preconditioner from a matrix using a Jacobi::Factory.
+     *
+     * @param factory  the factory to use to create the preconditoner
+     * @param system_matrix  the matrix this preconditioner should be created
+     *                       from
+     */
+    explicit Jacobi(const Factory *factory,
+                    std::shared_ptr<const LinOp> system_matrix)
+        : EnableLinOp<Jacobi>(factory->get_executor(),
+                              transpose(system_matrix->get_size())),
+          parameters_{factory->get_parameters()},
+          storage_scheme_{compute_storage_scheme(parameters_.max_block_size)},
+          num_blocks_{parameters_.block_pointers.get_num_elems() - 1},
+          blocks_(factory->get_executor(),
+                  storage_scheme_.compute_storage_space(
+                      parameters_.block_pointers.get_num_elems() - 1))
+    {
+        if (parameters_.max_block_size >= 32 ||
+            parameters_.max_block_size < 1) {
+            NOT_SUPPORTED(this);
+        }
+        parameters_.block_pointers.set_executor(this->get_executor());
+        parameters_.storage_optimization.block_wise.set_executor(
+            this->get_executor());
+        this->generate(lend(system_matrix));
+    }
+
+    /**
+     * Stride between two columns of a block (as number of elements).
+     *
+     * Should be a multiple of cache line size for best performance.
+     */
+    static constexpr size_type max_block_stride_ = 32;
+
+    /**
+     * Returns the smallest power of 2 at least as larger as the input.
+     *
+     * @param n  a number
+     *
+     * @return a power of two at least as large as `n`
+     */
+    static size_type get_larger_power(size_type n) noexcept
+    {
+        size_type res = 1;
+        while (res < n) res *= 2;
+        return res;
+    }
```

Comment:
Not sure. I didn't intend for these functions to be reusable, and I don't think the implementation quality is high enough to be reusable and high performance (e.g. it is not `constexpr` while it could be, it may be possible to implement them more efficiently using intrinsics in some cases, etc.). We would also pollute the namespace with functions whose usefulness is questionable.

Reply:
Made those a bit more general, and moved them [here](https://github.com/ginkgo-project/ginkgo/pull/167/files#diff-0745f99e4f9c2f0d10b5897a99edfc26).